newton.go

```diff
@@ -0,0 +1,143 @@
+//
+// If the Hessian matrix cannot be formed explicitly or if the computational
+// cost of its factorization is prohibitive, BFGS or L-BFGS quasi-Newton method
+// can be used instead.
+type Newton struct {
+	// LinesearchMethod is a method used for selecting suitable steps along the
+	// descent direction d. Steps should satisfy the Wolfe, Goldstein or Armijo
+	// conditions. If LinesearchMethod == nil, an appropriate default is
+	// chosen.
+	LinesearchMethod LinesearchMethod
+	// Increase is a factor by which a scalar tau is successively increased so
+	// that (H + tau*I) is positive definite. Larger values reduce the number
+	// of trial Hessian factorizations, but also reduce the second-order
+	// information in H.
+	// Increase must be greater than 1. If Increase is 0, it is defaulted to 5.
+	Increase float64
+
+	linesearch *Linesearch
+
+	hess *mat64.SymDense
+	chol *mat64.TriDense
+	tau  float64
+}
+
+func (n *Newton) Init(loc *Location, f *FunctionInfo, xNext []float64) (EvaluationType, IterationType, error) {
+	if n.Increase == 0 {
+		n.Increase = 5
+	}
+	if n.Increase <= 1 {
+		panic("optimize: Newton.Increase must be greater than 1")
+	}
+	if n.LinesearchMethod == nil {
+		n.LinesearchMethod = &Bisection{}
+	}
+	if n.linesearch == nil {
+		n.linesearch = &Linesearch{}
+	}
+	n.linesearch.Method = n.LinesearchMethod
+	n.linesearch.NextDirectioner = n
+
+	return n.linesearch.Init(loc, f, xNext)
+}
+
+func (n *Newton) Iterate(loc *Location, xNext []float64) (EvaluationType, IterationType, error) {
+	return n.linesearch.Iterate(loc, xNext)
+}
+
+func (n *Newton) InitDirection(loc *Location, dir []float64) (stepSize float64) {
+	dim := len(loc.X)
+	n.chol = resizeTriDense(n.chol, dim)
+	n.hess = resizeSymDense(n.hess, dim)
+	n.tau = 0
+	n.computeNextDir(loc, dir)
+	return 1
+}
+
+func (n *Newton) NextDirection(loc *Location, dir []float64) (stepSize float64) {
+	n.computeNextDir(loc, dir)
+	return 1
+}
+
+func (n *Newton) computeNextDir(loc *Location, dir []float64) {
+	dim := len(loc.X)
+	n.hess.CopySym(loc.Hessian)
+
+	// Find the smallest diagonal element of H.
+	minA := n.hess.At(0, 0)
+	for i := 1; i < dim; i++ {
+		a := n.hess.At(i, i)
+		if a < minA {
+			minA = a
+		}
+	}
+	if minA > 0 {
+		// All diagonal elements are positive, so try to factorize non-modified
+		// Hessian first.
+		n.tau = 0
+	} else if n.tau == 0 {
+		// Use the last tau that gave PD modified Hessian.
+		n.tau = -minA + 0.001
+	}
+
+	for {
+		if n.tau != 0 {
+			// Add a multiple of identity to the Hessian.
+			for i := 0; i < dim; i++ {
+				n.hess.SetSym(i, i, loc.Hessian.At(i, i)+n.tau)
+			}
+		}
+		// Try to apply the Cholesky decomposition.
+		pd := n.chol.Cholesky(n.hess, false)
+		if pd {
+			break
+		}
+		// Modified Hessian is not PD, so increase tau.
+		n.tau = math.Max(n.Increase*n.tau, 0.001)
+	}
+	d := mat64.NewDense(dim, 1, dir)
+	// Store the solution in d's backing array, dir.
```

Comment:
Add a TODO that this should use mat64.Vector SolveCholesky when it exists (or, if you're feeling charitable, implement it :) ). This way we can use Dtrsv instead of Dtrsm